Modules/arraymodule.c

```diff
@@ -130,11 +151,57 @@ enum machine_format_code {
 
 #define array_Check(op, state) PyObject_TypeCheck(op, state->ArrayType)
 
+static arraydata *
+arraydata_alloc(Py_ssize_t size, int itemsize)
+{
+    arraydata *data = (arraydata *)PyMem_Malloc(sizeof(arraydata) + size * itemsize);
+    if (data == NULL) {
+        return NULL;
+    }
+    data->allocated = size;
+    return data;
+}
+
+static void
+arraydata_free(arraydata *data, bool use_qsbr)
+{
+#ifdef Py_GIL_DISABLED
+    if (use_qsbr) {
+        _PyMem_FreeDelayed(data);
+    }
+    else {
+        PyMem_Free(data);
+    }
+#else
+    PyMem_Free(data);
+#endif
+}
+
+#ifndef Py_GIL_DISABLED
+
+static arraydata *
+arraydata_realloc(arraydata *data, Py_ssize_t size, int itemsize)
+{
+    data = (arraydata *)PyMem_Realloc(data, sizeof(arraydata) + size * itemsize);
+    if (data == NULL) {
+        return NULL;
+    }
+    data->allocated = size;
+    return data;
+}
+
+#endif
+
+static char *
```

Comment:
The compiler will probably optimize it anyway, but we might as well be explicit:
```suggestion
static inline char *
```